service-base/src/main/java/org/eclipse/hono/service/metric/DeviceConnectionTimeTracker.java

```diff
@@ -0,0 +1,167 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.metric;
+
+import io.vertx.core.TimeoutStream;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * TODO.
+ */
+public final class DeviceConnectionTimeTracker {
+
+    private static final long DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS = TimeUnit.SECONDS.toMillis(10);
+    protected final Logger log = LoggerFactory.getLogger(DeviceConnectionTimeTracker.class);
+    private final AtomicLong noOfDeviceConnections;
+    private final Consumer<Long> recorder;
+    private final AtomicReference<Instant> startInstant = new AtomicReference<>();
+    private final String tenantId;
+    private final TimeoutStream timer;
+
+    private DeviceConnectionTimeTracker(
+            final String tenantId, final Vertx vertx,
+            final long noOfDeviceConnections,
+            final Consumer<Long> recorder) {
+        this.tenantId = tenantId;
+        this.noOfDeviceConnections = new AtomicLong(noOfDeviceConnections);
+        this.recorder = recorder;
+        this.startInstant.set(Instant.now());
+        this.timer = vertx
+                .periodicStream(DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS)
+                .handler(id -> report(this.noOfDeviceConnections.get()));
+        log.trace("Stared a device connection time tracker for the tenant [{}].", tenantId);
+    }
+
+    /**
+     * Pauses the timer, calculates the device connection time and reports the result by invoking the consumer with the
+     * calculated value and then resumes the timer.
+     * TODO - more info to add.
+     * 
+     * @param noOfDeviceConnections The number of device connections.
+     * @return a reference to this for fluent use.
+     */
+    public DeviceConnectionTimeTracker updateNoOfDeviceConnections(final long noOfDeviceConnections) {
+        this.timer.pause();
+        report(this.noOfDeviceConnections.getAndSet(noOfDeviceConnections));
+        log.trace("Updated the number of device connections for the tenant [{}] to [{}]", tenantId,
+                noOfDeviceConnections);
+        if (noOfDeviceConnections == 0) {
+            this.timer.cancel();
+            log.trace("Stopped the device connection time tracker for the tenant [{}].", tenantId);
+            return null;
+        }
+        this.timer.resume();
+        return this;
+    }
+
+    private void report(final long noOfDeviceConnections) {
```

Comment:
FMPOV this should not need any argument as the current value to report can simply be read from the *noOfDeviceConnections* field, can't it?

Reply:
yes, I will update it.